### PR TITLE
OP-312 telemetry module

### PR DESCRIPTION
**Pull request by AndreiDodu**
See OP-312.

This PR is paired with https://github.com/informatici/openhospital-gui/pull/876
Mocked back-end Server available at https://github.com/AndreiDodu/telemetry-data-collector-server

### CODE REVIEW

File: src/main/java/org/isf/generaldata/GeneralData.java [hunk 1]
```diff
@@ -22,28 +22,22 @@
...
- * 20/12/2008 - isf bari - added patientextended
- * 01/01/2009 - Fabrizio - added OPDEXTENDED
- * 20/01/2009 - Chiara   - added attribute MATERNITYRESTARTINJUNE to reset progressive number of maternity ward
- * 25/02/2011 - Claudia  - added attribute MAINMENUALWAYSONTOP to handle main menu always on Top 
- * 01/05/2011 - Vito 	 - added attribute VIDEOMODULEENABLED to enable/disable video module
- * 10/08/2011 - Claudia  - added PATIENTVACCINEEXTENDED to show patient on Patient Vaccine 
- * 19/10/2011 - Mwithi   - GeneralData 2.0: catching exception on single property and assign DEFAULT value  
- * 29/12/2011 - Nicola   - added XMPPMODULEENABLED to enable/disable communication module
- * -------------------------------------------
+ * 12/2007 - isf bari - added resource bundle for internationalization 19/06/2008 - isf bari - added patientsheet jasper report name 20/12/2008 - isf bari -
```
[the remark this answers is not in the document]
Comment (AndreiDodu) [reply src/main/java/org/isf/generaldata/GeneralData.java]: corrected

File: src/main/java/org/isf/telemetry/model/Telemetry.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+	public void setHashCode(int hashCode) {
+		this.hashCode = hashCode;
+	}
+
+	public Map<String, Boolean> getConsentMap() {
+		if (this.consentData != null && !this.consentData.isEmpty()) {
+			Type type = new TypeToken<Map<String, Boolean>>() {
+			}.getType();
+			Gson gson = new Gson();
```
[the remark this answers is not in the document]
Comment (AndreiDodu) [reply src/main/java/org/isf/telemetry/model/Telemetry.java:132]: corrected

File: src/main/java/org/isf/utils/BooleanUtils.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.isf.utils;
+
+public class BooleanUtils {
+
+	public static boolean isTrue(Boolean bool) {
+		return bool != null && bool.booleanValue();
+	}
```
[the remark this answers is not in the document]
Comment (AndreiDodu) [reply src/main/java/org/isf/utils/BooleanUtils.java]: You are TRUE. When I am tired I make ugly things. Corrected.

File: src/main/java/org/isf/envdatacollector/collectors/CollectorsConst.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
+package org.isf.envdatacollector.collectors;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/envdatacollector/collectors/CollectorsConst.java:1]: Missing license header

Comment (AndreiDodu) [reply src/main/java/org/isf/envdatacollector/collectors/CollectorsConst.java:22]: corrected.

File: src/main/java/org/isf/patient/service/PatientIoOperations.java [hunk 5]
```diff
@@ -37,20 +37,10 @@
...
- * 26/08/2008 - claudio    - added birth date
- * modified age
- * 01/01/2009 - Fabrizio   - changed the calls to PAT_AGE fields to
- * return again an int type
- * 03/12/2009 - Alex       - added method for merge two patients history
- * ------------------------------------------
+ * ------------------------------------------ PatientIoOperations - dB operations for the patient entity ----------------------------------------- modification
+ * history 05/05/2005 - giacomo - first beta version 03/11/2006 - ross - added toString method. Gestione apici per nome, cognome, citta', indirizzo e note
+ * 11/08/2008 - alessandro - added father & mother's names 26/08/2008 - claudio - added birth date modified age 01/01/2009 - Fabrizio - changed the calls to
+ * PAT_AGE fields to return again an int type 03/12/2009 - Alex - added method for merge two patients history ------------------------------------------
```
Comment (mwithi) [inline src/main/java/org/isf/patient/service/PatientIoOperations.java:43]: no need to reformat all javadoc, especially those with a specific layout

File: src/main/java/org/isf/patient/service/PatientIoOperations.java [hunk 6]
```diff
@@ -197,8 +199,8 @@ public boolean deletePatient(Patient patient) throws OHServiceException {
 	}
 
 	/**
-	 * Method that check if a Patient is already present in the DB by his/her name
-	 * (the passed string 'name' should be a concatenation of firstName + " " + secondName
+	 * Method that check if a Patient is already present in the DB by his/her name (the passed string 'name' should be a concatenation of firstName + " " +
+	 * secondName
```
Comment (mwithi) [inline src/main/java/org/isf/patient/service/PatientIoOperations.java:203]: no need to reformat all javadoc, especially those with a specific layout

File: sql/step_69_disclose_accounting_admin_features.sql [hunk 7]
```diff
@@ -8,4 +8,4 @@ INSERT INTO MENUITEM (MNI_ID_A, MNI_BTN_LABEL, MNI_LABEL, MNI_TOOLTIP, MNI_SHORT
 INSERT INTO MENUITEM (MNI_ID_A, MNI_BTN_LABEL, MNI_LABEL, MNI_TOOLTIP, MNI_SHORTCUT, MNI_SUBMENU, MNI_CLASS, MNI_IS_SUBMENU, MNI_POSITION) VALUES ('editclosedbills', 'angal.menu.accounting.editclosedbills', 'angal.menu.accounting.editclosedbills', 'x', 'E', 'billsmanager', 'none', 'N', '3');
 
 INSERT INTO GROUPMENU (GM_UG_ID_A, GM_MNI_ID_A, GM_ACTIVE) VALUES ('admin', 'cashiersfilter', 1);
-INSERT INTO GROUPMENU (GM_UG_ID_A, GM_MNI_ID_A, GM_ACTIVE) VALUES ('admin', 'editclosedbills', 1);
+INSERT INTO GROUPMENU (GM_UG_ID_A, GM_MNI_ID_A, GM_ACTIVE) VALUES ('admin', 'editclosedbills', 1);
```
Comment (mwithi) [inline sql/step_69_disclose_accounting_admin_features.sql:11]: maybe unwanted change

File: src/main/java/org/isf/envdatacollector/AbstractDataCollector.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.isf.envdatacollector;
```
Comment (mwithi) [inline src/main/java/org/isf/envdatacollector/AbstractDataCollector.java:22]: I would put them all in `org.isf.telemetry.datacolletor`

File: src/main/java/org/isf/envdatacollector/collectors/ApplicationDataCollector.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+			String verMinor = props.getProperty("VER_MINOR");
+			String verRelease = props.getProperty("VER_RELEASE");
+			result.put(CollectorsConst.APP_VER_MAJOR, verMajor);
+			result.put(CollectorsConst.APP_VER_MINOR, verMinor);
+			result.put(CollectorsConst.APP_RELEASE, verRelease);
+		} catch (Exception e) {
+			LOGGER.error(e.toString());
+		}
+		return result;
+	}
```
Comment (mwithi) [inline src/main/java/org/isf/envdatacollector/collectors/ApplicationDataCollector.java:72]: should we reuse `/openhospital-core/src/main/java/org/isf/generaldata/Version.java`?

File: src/main/java/org/isf/envdatacollector/collectors/ApplicationDataCollector.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+			props.load(in);
+			in.close();
+		} else {
+			props.load(is);
+			is.close();
+		}
+		return props;
+	}
+
+}
```
Comment (mwithi) [inline src/main/java/org/isf/envdatacollector/collectors/ApplicationDataCollector.java:88]: not needed if we reuse `/openhospital-core/src/main/java/org/isf/generaldata/Version.java`

File: src/main/java/org/isf/envdatacollector/collectors/DBMSDataCollector.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+			String db = props.getProperty("jdbc.db");
+			String username = props.getProperty("jdbc.username");
+			String password = props.getProperty("jdbc.password");
+			String port = props.getProperty("jdbc.port");
+			if (port == null) {
+				port = String.valueOf(MYSQL_DEFAULT_PORT);
+			}
+			StringBuilder sbURL = new StringBuilder();
+			sbURL.append(url);
+			sbURL.append("//");
```
Comment (mwithi) [inline src/main/java/org/isf/envdatacollector/collectors/DBMSDataCollector.java:82]: Should we reuse `/openhospital-core/src/main/java/org/isf/utils/db/DbSingleConn.java` ?

File: src/main/java/org/isf/envdatacollector/collectors/DBMSDataCollector.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+		if (is == null) {
+			FileInputStream in = new FileInputStream("rsc/database.properties");
+			props.load(in);
+			in.close();
+		} else {
+			props.load(is);
+			is.close();
+		}
+		return props;
+	}
```
Comment (mwithi) [inline src/main/java/org/isf/envdatacollector/collectors/DBMSDataCollector.java:127]: Not needed if we reuse `/openhospital-core/src/main/java/org/isf/utils/db/DbSingleConn.java`

File: src/main/java/org/isf/envdatacollector/collectors/OpenHospitalDataCollector.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+	public Map<String, String> retrieveData() {
+		LOGGER.debug("Collecting Open Hospital data...");
+		Map<String, String> result = new HashMap<>();
+		try {
+			result.put(CollectorsConst.OH_ACTIVE_PATIENTS, String.valueOf(patientIoOperations.countAllActivePatients()));
+			result.put(CollectorsConst.OH_ACTIVE_USERS, String.valueOf(this.menuIoOperations.countAllActive()));
+			// TODO retrieve number of beds
+			// result.put("oh_beds", );
+			// TODO retrieve number of departments
+			// result.put("oh_departments", );
```
Comment (mwithi) [inline src/main/java/org/isf/envdatacollector/collectors/OpenHospitalDataCollector.java:70]: To use `/openhospital-core/src/main/java/org/isf/ward/manager/WardBrowserManager.java`

File: src/main/java/org/isf/envdatacollector/collectors/TimeDataCollector.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+	@Override
+	public String getDescription() {
+		return "Time information (ex. last used timestamp)";
+	}
+
+	@Override
+	public Map<String, String> retrieveData() {
+		LOGGER.debug("Collecting Time data...");
+		Map<String, String> result = new HashMap<>();
+		result.put(CollectorsConst.TIME_LAST_USED, (new Date()).toString());
```
Comment (mwithi) [inline src/main/java/org/isf/envdatacollector/collectors/TimeDataCollector.java:55]: I think here we should collect last date in the DB records, among all tables... Maybe the most recent `PAT_CREATED_DATE` or `OPD_CREATED_DATE` will be enough

File: src/main/java/org/isf/utils/ExceptionUtils.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.isf.utils;
+
+public class ExceptionUtils {
```
Comment (mwithi) [inline src/main/java/org/isf/utils/ExceptionUtils.java:24]: What is this class for?

File: src/main/java/org/isf/utils/time/DateUtils.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+package org.isf.utils.time;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+public class DateUtils {
+
+	private static final String DATE_FORMAT = "yyyy-MM-dd";
+
+	public static boolean isNotToday(Date date) {
```
Comment (mwithi) [inline src/main/java/org/isf/utils/time/DateUtils.java:31]: possible to reuse `org.isf.utils.time.TimeTools.isSameDay(Date, Date)` ?

File: src/main/java/org/isf/patient/service/PatientIoOperations.java (hunk 5 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/patient/service/PatientIoOperations.java]: corrected

File: sql/step_69_disclose_accounting_admin_features.sql (hunk 7 above)
Comment (AndreiDodu) [reply sql/step_69_disclose_accounting_admin_features.sql]: corrected

File: src/main/java/org/isf/envdatacollector/AbstractDataCollector.java (hunk 8 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/envdatacollector/AbstractDataCollector.java]: Corrected.

File: src/main/java/org/isf/envdatacollector/collectors/ApplicationDataCollector.java (hunk 9 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/envdatacollector/collectors/ApplicationDataCollector.java]: good idea :). Done.

File: src/main/java/org/isf/envdatacollector/collectors/ApplicationDataCollector.java (hunk 10 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/envdatacollector/collectors/ApplicationDataCollector.java]: done

File: src/main/java/org/isf/envdatacollector/collectors/DBMSDataCollector.java (hunk 12 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/envdatacollector/collectors/DBMSDataCollector.java]: ok

File: src/main/java/org/isf/envdatacollector/collectors/TimeDataCollector.java (hunk 14 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/envdatacollector/collectors/TimeDataCollector.java]: done (used OPD_CREATED_DATE)

File: src/main/java/org/isf/utils/time/DateUtils.java (hunk 16 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/utils/time/DateUtils.java]: corrected

File: src/main/java/org/isf/patient/service/PatientIoOperations.java (hunk 6 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/patient/service/PatientIoOperations.java]: corrected

File: src/main/java/org/isf/envdatacollector/collectors/DBMSDataCollector.java (hunk 11 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/envdatacollector/collectors/DBMSDataCollector.java]: done

File: src/main/java/org/isf/utils/ExceptionUtils.java (hunk 15 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/utils/ExceptionUtils.java:27]: I need it to retrieve informations about exceptions in OH. Currently ExceptionUtils helps to retrieve  exception stacktrace that then is passed to the logger.

File: src/main/java/org/isf/envdatacollector/collectors/OpenHospitalDataCollector.java (hunk 13 above)
Comment (AndreiDodu) [reply src/main/java/org/isf/envdatacollector/collectors/OpenHospitalDataCollector.java]: done

File: src/main/java/org/isf/telemetry/envdatacollector/collectors/DBMSDataCollector.java [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+		Connection con = null;
+		try {
+			con = DbSingleConn.getConnection();
+			DatabaseMetaData dbmd = con.getMetaData();
+			result.put(CollectorsConst.DBMS_DRIVER_NAME, dbmd.getDriverName());
+			result.put(CollectorsConst.DBMS_DRIVER_VERSION, dbmd.getDriverVersion());
+			result.put(CollectorsConst.DBMS_USERNAME, dbmd.getUserName());
+			result.put(CollectorsConst.DBMS_PRODUCT_NAME, dbmd.getDatabaseProductName());
+			result.put(CollectorsConst.DBMS_PRODUCT_VERSION, dbmd.getDatabaseProductVersion());
+			return result;
```
Comment (mwithi) [inline src/main/java/org/isf/telemetry/envdatacollector/collectors/DBMSDataCollector.java:75]: Maybe there is a way to get those informations from the Hibernate existing connection ?
Anyway we could "upgrade" `DbSingleConn` to use actual keys in `database.properties.dist `

Comment (AndreiDodu) [reply src/main/java/org/isf/telemetry/envdatacollector/collectors/DBMSDataCollector.java:75]: I made only refactoring. Now it loads db connection configuration from the same file.

**Pull request merged**